server/configuration-validation.js

```diff
@@ -1,125 +1,192 @@
 const { isObject } = require('./utils')
 
-/**
- * Validates the plugin configuration provided in `plugins/config.js` of the users plugin configuration.
- * Modifies the value of config on place.
- *
- * @param  {object} config - configurations
- */
-function validateConfiguration(config) {
-  if (config === undefined) {
-    return
-  }
+function CollectionOptions(name, configuration) {
+  const log = strapi.log // has to be inside a scope
+  const {
+    indexName,
+    transformEntry,
+    filterEntry,
+    settings,
+    populateEntryRule,
+    ...excedent
+  } = configuration
+  const options = {}
 
-  if (!isObject(config)) {
-    strapi.log.error(
-      'The `config` field in the Meilisearch plugin configuration must be of type object'
-    )
-    config = {}
-  }
-  const { host, apiKey, ...collections } = config
-
-  // Validate the `host` parameter
-  if ((host !== undefined && typeof host !== 'string') || config.host === '') {
-    strapi.log.error(
-      '`host` should be a non-empty string in Meilisearch configuration'
-    )
-    delete config.host
-  }
+  return {
+    validateConfiguration() {
```

Comment:
Could be `validateCollection` or `validateCollectionObject`

Reply:
What do you think about:

- Change the first occurrence of `validateConfiguration` which is called by the [`strapi-server.js`](https://github.com/meilisearch/strapi-plugin-meilisearch/blob/baaa9d1f31fa4dc163458573afa41fd3f6b96734/strapi-server.js#L17) with `validatePluginConfiguration` configuration

https://github.com/meilisearch/strapi-plugin-meilisearch/blob/baaa9d1f31fa4dc163458573afa41fd3f6b96734/strapi-server.js#L16-L18

- Check the format of the configuration before using `PluginOptions`

```js
function validatePluginConfiguration(configuration) {
  // If no configuration, return
  if (configuration === undefined) {
    return
  } else if (configuration !== undefined && !isObject(configuration)) {
    log.error(
      'The `config` field in the Meilisearch plugin configuration must be an object'
    )
    return
  }

  const options = PluginOptions(configuration)
    .validateApiKey()
    .validateHost()
    .validateCollections()
    .get()

  Object.assign(configuration, options)
```
- Same for `validateCollections`

```js
     for (const collection in collections) {
        if (!isObject(collections[collection])) {
          log.error(`The collection "${collection}" should be of type object`)
          options[collection] = {}
        } else {
          options[collection] = CollectionConfig(
            collection,
            collections[collection]
          )
            .validateIndexName()
            .validateFilterEntry()
            .validateTransformEntry()
            .validateMeilisearchSettings()
            .validatePopulateEntryRule()
            .validateNoInvalidKeys()
            .get()
        }
```

Reply:
I think it's a lot clearer!